tools/PI/DevHome.PI/Helpers/WindowHelper.cs

```diff
@@ -337,6 +339,69 @@ public static async Task<SoftwareBitmapSource> GetSoftwareBitmapSourceFromSoftwa
         return softwareBitmapSource;
     }
 
+    internal static BitmapImage? GetBitmapImageFromFile(string filePath)
+    {
+        BitmapImage? bitmapImage = null;
+        try
+        {
+            // Creating a BitmapImage must be done on the UI thread.
+            bitmapImage = new BitmapImage();
+            var icon = Icon.ExtractAssociatedIcon(filePath);
+            if (icon is not null)
+            {
+                using var bitmap = icon.ToBitmap();

```

Comment:
>u [](http://example.com/codeflow?start=16&length=1)

Look up above in GetSoftwareBitmapFromExecutable(). No need to write the icon to a memory stream...

Reply:
In fact, does the above function do what you need?

Reply:
Well, GetSoftwareBitmapFromExecutable uses Icon.ExtractAssociatedIcon, which doesn't work on all file types (works best on executables) (I did try this first). It's also unfortunate that we need different image types for the list, toolbar and menu. Between Icon, IconImage, Bitmap, BitmapImage, SoftwareBitmap, SoftwareBitmapSource etc, etc - you'd think there'd be easy transforms from one to the other, but no.